gobblin-runtime/src/main/java/org/apache/gobblin/runtime/kafka/HighLevelConsumer.java

```diff
@@ -237,13 +257,16 @@ protected void startUp() {
   private void consume() {
     try {
       Iterator<KafkaConsumerRecord> itr = gobblinKafkaConsumerClient.consume();
+      // TODO: we may be committing too early and only want to commit after process messages
       if(!enableAutoCommit) {
         commitOffsets();
       }
       while (itr.hasNext()) {
         KafkaConsumerRecord record = itr.next();
         int idx = record.getPartition() % numThreads;
         queues[idx].put(record);
+        // Increment queue size metric
+        updateQueueSizes(idx);
```

Comment:
this is a pretty fundamental change to put into the fast path!  we gotta be certain that no critical use is adversely affected by perf regression.

I see `ConcurrentLinkedQueue::size` may not be `O(1)`, but I can't find a statement on `LinkedBlockingQueue::size`...

(it's possible it IS `O(1)`, but an approximation or else exact, but at the expense of coordination overhead.)

Reply:
Alternatively, I was thinking we can have a background thread update the size every minute (or some other interval). That will move this out of the fast path. Alternatively we can call this method in an async thread.